### ISSUE TITLE
in-place format

### ISSUE THREAD

**Issue by NickeZ** [kkinnear/zprint#159]
Hi, does zprint have support for in-place formatting? I end up with empty files if I do `zprint < file.clj > file.clj`. Thanks

**Comment by kkinnear** [comment]
Good question!  Turns out that, while zprint does read the input before writing the output, that doesn't help when pointing stdin and stdout to the same file.  What I understand to be happening is that once stdout points to a file, the shell effectively clears it and then when stdin reads the file -- there is nothing in it.  Can't really work around that with zprint, as the shell has already done the work by the time zprint runs.

Given that you clearly would like this to work, I'm working on a new swtich which would let zprint do the in-place formatting:
```
zprint -r file.clj

zprint --replace file.clj
```
If you wanted an options-map, it would look like this:
```
zprint -r file.clj '{:width 90 :style :justified}'
```
I think that would meet people's needs for in-place formatting.  Let me know if that would work for you. Thanks!

**Comment by frou** [comment]
There's been a Unix command around for a while to handle this situation called `sponge`. The metaphor is that it soaks up the entirety of its standard input, and *then* wrings that out into a file.

used like:

`zprint < file.clj | sponge file.clj`

But I don't think it's very well known (and is distributed bundled with other stuff in [moreutils](https://repology.org/project/moreutils/versions)) so probably a bit too annoying to rely on!

**Comment by agriffis** [comment]
If you're going to do that, this works as well:

```
(rm -f file.clj; zprint > file.clj) < file.clj
```

because it opens the file, unlinks it, and then outputs to a new file of the same name.

The problem with that approach, and with `sponge`, is that it depends on everything going well with the formatter. If the formatter fails, for example it's not in the PATH, or the command-line is wrong, then the source file will be truncated.

You can do more wrapping to avoid that outcome, but eventually it's just easier to let the formatter handle it. 😄

**Comment by agriffis** [comment]
Just for fun, here is the fuller version, a little bit golfed:

```bash
# inplace command args -- files
inplace() {
  declare f; declare -a cmd
  while [[ ${1---} != -- ]]; do cmd+=( "$1" ); shift; done
  shift
  for f; do
    "${cmd[@]}" < "$f" > "$f.tmp" || return
    mv "$f.tmp" "$f" || return
  done
}
```

to be used like this:

```bash
inplace zprint args -- file1.clj file2.clj file3.clj
```

**Comment by frou** [comment]
@agriffis Nice shell-fu. The idea with `sponge` is that if the first command fails to parse the file, then it shouldn't write anything to stdout. That gets noticed and so the file doesn't get modified either. At least that's [what I did](https://github.com/frou/sponge/blob/master/main.go#L46) in a Go port.

Anyway, this is somewhat off-topic I guess. It's a welcome thing for `zprint` to support in-place formatting natively!

**Comment by NickeZ** [comment]
> Good question! Turns out that, while zprint does read the input before writing the output, that doesn't help when pointing stdin and stdout to the same file. What I understand to be happening is that once stdout points to a file, the shell effectively clears it and then when stdin reads the file -- there is nothing in it. Can't really work around that with zprint, as the shell has already done the work by the time zprint runs.
> 
> Given that you clearly would like this to work, I'm working on a new swtich which would let zprint do the in-place formatting:
> 
> ```
> zprint -r file.clj
> 
> zprint --replace file.clj
> ```
> 
> If you wanted an options-map, it would look like this:
> 
> ```
> zprint -r file.clj '{:width 90 :style :justified}'
> ```
> 
> I think that would meet people's needs for in-place formatting. Let me know if that would work for you. Thanks!

Yea this sounds like what I want :), just for reference `-i` is commonly used for in-place, like in `sed`. But you are free to choose whatever you prefer!

**Comment by kkinnear** [comment]
So, about `--in-place` -- I'm always up for a good precedent! That said, I have been unable to find any common utilities that use `--in-place` other than `sed` and a rather obscure Java asm annotation tool.  Maybe I'm just not looking in the right places?  Do you know of any others off the top of your head?  

I'm concerned that people will mistake `-i` for `--input`, and be unhappy when it also means `--output` as well.  Thanks!

**Comment by agriffis** [comment]
[prettier](https://prettier.io/docs/en/cli.html) uses `--write`

[black](https://black.readthedocs.io/en/stable/installation_and_usage.html) assumes in-place unless other options override

[gofmt](https://golang.org/cmd/gofmt/) uses `-w`

[rustfmt](https://github.com/rust-lang/rustfmt#user-content-tips) writes by default, can redirect to stdout with `--emit stdout`

For what it's worth, sed's `--in-place` is a bit of a misnomer. It doesn't actually write the file in-place, rather it unlinks (or renames) the existing file and writes a new one:

```
$ echo hi > foo
$ ln foo bar
$ sed -i s/i/ello/ foo
$ cat foo
hello
$ cat bar
hi
```

I don't really have a preference or suggestion, but there are some inputs for discussion...

**Comment by agriffis** [comment]
[clang-format](http://clang.llvm.org/docs/ClangFormat.html) works the most like sed with `-i`:

> If no arguments are specified, it formats the code from standard input and writes the result to the standard output. If \<file>s are given, it reformats the files. If -i is specified together with \<file>s, the files are edited in-place. Otherwise, the result is written to the standard output.

**Comment by frou** [comment]
`ocamlformat` has:

```
-i, --inplace
        Format in-place, overwriting input file(s).
```

`-w` `--write` seems saner though.

I don't think people just try vaguely conventional flags blindy. We look at the `--help`, surely.

**Comment by kkinnear** [comment]
Thanks to you all for the input.  Based on it, my current plan is to continue to read from stdin and write to stdout, unless the `-w` or `--write` option is specified.  If it given, the file specified will be read and then rewritten with the reformatted source.

So these would both work:
```
zprint '{:width 90}' -w file.clj

zprint -w file.clj
```
This would not work:
```
zprint -w file.clj '{:width 90}'
```
largely because this would be very strange:
```
zprint -w *.clj '{:width 90}'
```
but these would both work:
```
zprint -w *.clj
zprint '{:width 90}' -w *.clj
```
which would be a significant change, but one that I think people might welcome.

It would open up some questions, not least of which is -- what happens when one of the files fails to format -- what happens to the files later in the list of `*.clj` files?  

My thinking is to leave the file that failed to format unaltered (or rewrite the original information in it, whichever seems most straightforward), output some information to stderr, and move on to process the rest of the files.  I think that the error status in that case should be 1 not 0.

Any opinions about this?

**Comment by agriffis** [comment]
@kkinnear Is this issue holding up the release that would fix #56? I know that for me, personally, this issue is a nice to have—especially since there are workarounds—but that issue effectively prevents me from using zprint for a client project.

**Comment by kkinnear** [comment]
Well, yes, it is.  I have a solution to this issue (#159) coded up.  Given what you said, if I can't get it to work in a day or so, I'll cut a new release without it and do it next release.  Hopefully that will work for you!

**Comment by agriffis** [comment]
Sounds great, thanks! Really appreciate all your work on zprint, Kim!

**Comment by frou** [comment]
> Any opinions about this?

Everything you outlined there feels intuitive to me. Seems a good plan.

**Comment by kkinnear** [comment]
Available in 1.0.2.

Thanks @NickeZ for asking for this.  I don't know why I didn't think about doing this myself, but it is a particularly useful feature and makes `lein zprint` not all that useful -- or necessary.  It is also a lot faster than `lein zprint`.

Thanks also to @frou and  @agriffis for your help and suggestions for this.  I would have named it differently, and I'm glad I didn't.

This is really useful -- I'm already using it myself!

I will be adding capability to this in the next release, so that more of the other switches can be used with it (it is standalone right now), and so that there is a way to (optionally) get zprint to output information about what it is doing along the way.

I really appreciate all of the interest and opinions.  Gets kind of lonely sometimes, thanks for the input!
